chain-signatures/node/src/protocol/sync/mod.rs

```diff
@@ -0,0 +1,630 @@
+const REQUEST_UPDATE_TIMEOUT: Duration = Duration::from_millis(50);
+/// How long for a synchronous [`ProtocolRequest`] to get back a [`ProtocolResponse`] before timeout.
+const REQUEST_PROTOCOL_TIMEOUT: Duration = Duration::from_millis(50);
+/// The maximum number of protocol requests that can be queued. This will likely never get to
+/// this upper bound just purely based on our ProtocolConfig.
+const MAX_PROTOCOL_REQUESTS: usize = 128 * 1024;
+/// The maximum number of update requests that can be queued. This is pretty much just
+/// based on the number of participants in the network. If we have 1024 participants then
+/// our issue will more than likely not be the channel size.
+const MAX_SYNC_UPDATE_REQUESTS: usize = 1024;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SyncUpdate {
+    triples: Vec<TripleId>,
+    presignatures: Vec<PresignatureId>,
+}
+
+impl SyncUpdate {
+    pub fn empty() -> Self {
+        Self {
+            triples: Vec::new(),
+            presignatures: Vec::new(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.triples.is_empty() && self.presignatures.is_empty()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SyncView {
+    triples: HashSet<TripleId>,
+    presignatures: HashSet<PresignatureId>,
+}
+
+impl SyncView {
+    fn empty() -> Self {
+        Self {
+            triples: HashSet::new(),
+            presignatures: HashSet::new(),
+        }
+    }
+}
+
+struct SyncCache {
+    me: Participant,
+    /// The set of self owned triples seen by both us and other participants.
+    owned_triples: HashMap<TripleId, Vec<Participant>>,
+    /// The set of self owned presignatures seen by both us and other participants.
+    owned_presignatures: HashMap<PresignatureId, Vec<Participant>>,
+}
+
+impl SyncCache {
+    fn new(me: Participant) -> Self {
+        Self {
+            me,
+            owned_triples: HashMap::new(),
+            owned_presignatures: HashMap::new(),
+        }
+    }
+
+    fn update(
+        &mut self,
+        update: SyncUpdate,
+        views: Vec<(Participant, SyncView)>,
+        threshold: usize,
+    ) {
+        // Clear the cache before updating it since we are only taking a temporary view of who has a share
+        // of the triples/presignatures. This is so storage for each can be updated and we can adapt to it.
+        self.clear();
+
+        // Update the cache with our own info:
+        for id in update.triples {
+            let entry = self.owned_triples.entry(id).or_default();
+            entry.push(self.me);
+        }
+        for id in update.presignatures {
+            let entry = self.owned_presignatures.entry(id).or_default();
+            entry.push(self.me);
+        }
+
+        // Update the cache with the info from other participants:
+        for (p, view) in views {
+            for triple in view.triples {
+                let entry = self.owned_triples.entry(triple).or_default();
```

Comment:
(for future PRs, maybe add TODO for now)

When this entry doesn't exist, yet, it means the peer holds a triple that we no longer have. There should be some mechanism to have them delete it.

It could be an explicit message, requesting a deletion of all IDs we find here to be obsolete.

Or it could be that we do nothing here and instead send a list of which IDs we currently hold on each sync and then the peer can delete it by iterating through all foreign IDs and removing what is obsolete then.